src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
+        /// <param name="waitHandleToUse">
+        /// WaitHandle to use for waiting on the job object.
+        /// </param>
+        internal void WaitOne(ManualResetEvent waitHandleToUse)
         {
-            ManualResetEvent emptyJobAutoEvent = (ManualResetEvent)stateInfo;
-            int dwSize = 0;
-            const int JOB_OBJECT_BASIC_PROCESS_ID_LIST = 3;
-            JOBOBJECT_BASIC_PROCESS_ID_LIST JobList = new();
+            if (this._completionPortHandle == nint.Zero)
+            {
+                return;
+            }
+
+            const int INFINITE = -1;
+            int completionCode = 0;
+            do
+            {
+                Interop.Windows.GetQueuedCompletionStatus(
+                    this._completionPortHandle,
+                    out completionCode,
+                    out _,
+                    out _,
+                    INFINITE);
+            }
+            while (completionCode != Interop.Windows.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO);
+        }
+
+        private bool InitializeJob()
+        {
+            if (this._initStatus is not null)
+            {
+                return (bool)this._initStatus;
+            }
 
-            dwSize = Marshal.SizeOf(JobList);
-            if (NativeMethods.QueryInformationJobObject(_jobObjectHandle,
-                JOB_OBJECT_BASIC_PROCESS_ID_LIST,
-                ref JobList, dwSize, IntPtr.Zero))
+            if (this._jobObjectHandle == nint.Zero)
             {
-                if (JobList.NumberOfAssignedProcess == 0)
+                this._jobObjectHandle = Interop.Windows.CreateJobObject(nint.Zero, nint.Zero);
+                if (this._jobObjectHandle == nint.Zero)
                 {
-                    emptyJobAutoEvent.Set();
+                    this._initStatus = false;
+                    return false;
                 }
             }
+
+            if (this._completionPortHandle == nint.Zero)
+            {
+                this._completionPortHandle = Interop.Windows.CreateIoCompletionPort(
+                    -1,
+                    nint.Zero,
+                    nint.Zero,
+                    1);
+                if (this._completionPortHandle == nint.Zero)
+                {
+                    this._initStatus = false;
+                    return false;
+                }
+            }
+
+            var completionPort = new Interop.Windows.JOBOBJECT_ASSOCIATE_COMPLETION_PORT()
+            {
+                CompletionKey = this._jobObjectHandle,
+                CompletionPort = this._completionPortHandle,
+            };
+
+            unsafe
+            {
+                this._initStatus = Interop.Windows.SetInformationJobObject(
+                    this._jobObjectHandle,
+                    Interop.Windows.JobObjectAssociateCompletionPortInformation,
+                    (nint)(&completionPort),
+                    Marshal.SizeOf<Interop.Windows.JOBOBJECT_ASSOCIATE_COMPLETION_PORT>());
+            }
```

Comment:
Please move unsafe code to Interop class. We can create new helper method there and call it here.

Reply:
Any common location to put these wrapper functions?

Reply:
Near p/invoke definition. One example FindFirstFile.cs or FindExecutable.cs